### ISSUE TITLE
Returning String from rs to cs via service pattern

### ISSUE THREAD

**Issue by wgershwitz** [ralfbiedert/interoptopus#20]
Thank you for this awesome library!

It was really easy to get started with the provided examples, but I'm having hard time to return String (AsciiPointer)
*more than once* from Rust to CS via the Service pattern. 

With the following snippet I managed to return the string once, but the second call is causing the program to crash:
```
use crate::result::{Error, FFIError};
use interoptopus::{ffi_service, ffi_service_ctor, ffi_service_method, ffi_type};
use interoptopus::patterns::string::AsciiPointer;
use std::ffi::CString;

#[ffi_type(opaque)]
#[derive(Default)]
pub struct Sample {
    pub number: usize,
    name: CString,
}


#[ffi_service(error = "FFIError", prefix = "sample_")]
impl Sample {
    #[ffi_service_ctor]
    pub fn new_with(number: u32) -> Result<Self, Error> {
        Ok(Self {
            number: number as usize,
            name: CString::new(format!("HELLO_{}", number)).unwrap(),
        })
    }

    #[ffi_service_method(direct)]
    #[allow(non_snake_case)]
    pub fn Name(&self) -> AsciiPointer {
        // FIXME: Name can only be called once. second invocation causes crash.
        AsciiPointer::from_cstr(&self.name)
    }
}

interoptopus::inventory!(
    my_inventory, 
    [], 
    [], 
    [], 
    [Sample]);
```

My theory is that the `Sample.CString` gets freed after `AsciiPointer::from_cstr` is returned.
I tried to solve this by introducing new lifetime for name, but didn't succeed: always ended up having obscure
error coming from the ffi_service macro.

Any suggestion / help would be greatly appreciated!

**Comment by ralfbiedert** [comment]
Ha, good catch, can reproduce. I'm not entirely sure why this happens but I assume C# must do something with the pointer it "isn't supposed to", like trying to dealloc the pointer. This is bad and needs a fix. Investigating.

**Comment by wgershwitz** [comment]
Thank you for taking a look here!

**Comment by wgershwitz** [comment]
this might be relevant:
https://doc.rust-lang.org/std/ffi/struct.CStr.html#method.as_ptr

**Comment by ralfbiedert** [comment]
Thanks. I think it's unrelated to Rust's ptr. I've just read 

https://limbioliong.wordpress.com/2011/06/16/returning-strings-from-a-c-api/

and believe the underlying issue is C# thinks it's allowed to own the pointer's underlying memory and dealloc it. Given how Rust's alloc works I don't think there is any way on the Rust side how this could be solved (in Rust only exactly the matching allocator / deallocators must be used, while C# would deallocate this with 'whatever' deallocator).

In other words, this either must be

- a C# attribute (I'm still trying to find ) on the method, or
- not returning `string` types at all, but something like `InteropString` instead, although then the question will be how to soundly get a `string` from that one

**Comment by wgershwitz** [comment]
Thanks for the investigation. 
I'm a little bit confused about free, as other FFI doc suggests that freeing the resource should be done in rust:
http://jakegoulding.com/rust-ffi-omnibus/string_return/

Not having `string` is kind of a deal-breaker for me: I would like to implement an interface which should be string.
I could of course make a facade, but it would be nicer to have this ability.

**Comment by ralfbiedert** [comment]
>  I'm a little bit confused about free, as other FFI doc suggests that freeing the resource should be done in rust:

What basically happend:
> Rust: _"Here is a pointer to ASCII data I own, you may read that."_
> C#: _"Nice pointer you got there, let me make a `string` out of that one and free it for you"_

In other words, declaring the interop return type `string` makes C# think 1) it has ownership and 2) tries to dealloc the string  afterwards. Both are bad and C# should do neither. 

The link you posted is one way around it, you create a special class and tell it that when it destructs it should invoke a special C-level destructor. This is what I thought to do at first, but the ugly part is you'd have to call `service.GetName().ToString()` as above. 

However, it turns out we can just internally return an `IntPtr` and then call `Marshal.PtrToStringAnsi` at the right place, which is both more lightweight and C-friendly compatible. 

Long story short, I think it's fixed in 0.12, available in ~5 minutes

**Comment by ralfbiedert** [comment]
Ah, on a side note, you should not do `allow(non_snake_case)` in here:

```
    #[ffi_service_method(direct)]
    #[allow(non_snake_case)]
    pub fn Name(&self) -> AsciiPointer {
        AsciiPointer::from_cstr(&self.name)
    }
```

Services (should, sans bugs) translate methods to proper capitalization, i.e., Rust `name()` becomes `Name()` in C# and `service_name` in C.
